### DISCUSSION TITLE
[Invidious] HMAC_KEY config

### DISCUSSION THREAD

**Discussion by iBicha** [runtipi/runtipi-appstore#2127] [Q&A] [answered]
I was investigating an [issue](https://github.com/iv-org/invidious/issues/3540), and I saw it's very important to configure HMAC_KEY, to prevent access tokens from failing validation.

This has been added to the [docs](https://github.com/iv-org/documentation/commit/aee6d06eb827db59ce2045bcd8d196724b1b2841).

The hmac_key can be set into the docker compose file. But it is supposed to be a random number.

Should this be randomly generated on first start?

If it is hardcoded to a certain value, then that's a security issue, especially for people exposing Invidious.

Perhaps generating a value, caching it and setting it to an env var makes sense here.

**Comment by SamantazFox** [comment]
you can use the environment variable `INVIDIOUS_HMAC_KEY` to set said key.
If you want, this can be combined a `.env` file and a cron job to regularly update the value (e.g every 3 months)

**Comment by iBicha** [comment]
Nice! 
Here's some additional config items for a properly working instance in a local network environment:
- Setting the hmac_key
- Setting the domain (this is needed so that the api return valid results)
- Setting the external_port to 8095, which is the ${APP_PORT} (if this is not done, /api calls would return results with port 3000)

I guess these configs are also needed if users choose to "expose app" option

**Comment by nicotsx** [comment]
Hello, thanks for your input! We can definitely generate this randomly. This mechanism is already in place for db passwords. However since this needs a specific format, it would need a new development + release

**Comment by iBicha** [comment]
@meienberger #796 brought my attention to this, so I tried to fix it as well
I've tried setting these
```
        external_port: ${APP_PORT}
        domain: ${APP_DOMAIN}
```
But urls returned from the api ends up looking like this
```
http://192.168.1.119:8095:8095/path...
```
Invidious expects domain to not contain the port.
What would be the right variable to use here?

**Comment by nicotsx** [comment, marked as answer]
I can add a new variable `APP_HOST` to fulfill your requirement and release a new patch version. Currently this value does not exist

**Comment by iBicha** [comment]
> I can add a new variable `APP_HOST` to fulfill your requirement and release a new patch version. Currently this value does not exist

That would be nice! 
Also, should apps require a minimum version of Tipi? I'm wondering what would happen if people updated apps, but not tipi.

**Comment by nicotsx** [comment]
Great point, adding this as well as an option in `config.json`
